app/cruds/cruds_elocaps.py

```diff
@@ -0,0 +1,261 @@
+        select(models_elocaps.Game)
+        .where(models_elocaps.Game.id == game_id)
+        .options(
+            selectinload(models_elocaps.Game.game_players)
+            .selectinload(models_elocaps.GamePlayer.player)
+            .selectinload(models_elocaps.Player.user)
+        )
+    )
+    return result.scalars().first()
+
+
+async def register_game(
+    db: AsyncSession,
+    game: models_elocaps.Game,
+    game_players_info: list[schemas_elocaps.GamePlayerBase],
+) -> None:
+    try:
+        db.add(game)
+        await db.commit()
+        for i in game_players_info:
+            player = (
+                (
+                    await db.execute(
+                        select(models_elocaps.Player).where(
+                            (models_elocaps.Player.user_id == i.user_id)
+                            & (models_elocaps.Player.mode == game.mode)
+                        )
+                    )
+                )
+                .scalars()
+                .first()
+            )
+            if not player:
+                player = models_elocaps.Player(user_id=i.user_id, mode=game.mode)
+                db.add(player)
+                await db.commit()
+            game_player = models_elocaps.GamePlayer(
+                game_id=game.id, player_id=player.id, team=i.team, score=i.score
+            )
+            db.add(game_player)
+        await db.commit()
+        complete_game = (
+            (
+                await db.execute(
+                    select(models_elocaps.Game)
+                    .where(models_elocaps.Game.id == game.id)
+                    .options(
+                        selectinload(models_elocaps.Game.game_players).selectinload(
+                            models_elocaps.GamePlayer.player
+                        )
+                    )
+                )
+            )
+            .scalars()
+            .first()
+        )
+        if complete_game is None:
+            raise Exception("The game that was there just disappeared, we are screwed")
+        else:
+            game = complete_game
+        for game_player in game.game_players:
+            team = game_player.team
+            game_player.elo_gain = round(
+                compute_elo_gain(
+                    game.get_team_score(team),
+                    20,
+                    game.get_team_elo(team),
+                    game.get_team_elo(-team + 3),
+                )
+            )
+        await db.commit()
+    except IntegrityError as error:
+        await db.rollback()
+        await db.delete(game)
+        await db.commit()
+        raise ValueError(error)
+
+
+async def get_waiting_games(
+    db: AsyncSession, user_id: str
+) -> list[models_elocaps.Game]:
+    result = await db.execute(
+        select(models_elocaps.GamePlayer)
+        .join(models_elocaps.Player)
+        .where(
+            (models_elocaps.Player.user_id == user_id)
+            & ~models_elocaps.GamePlayer.has_confirmed
```

Comment:
I'm not sure the tilde operator behaviour is obvious for our devs. If I understand correctly the objective here is to to a negation, wouldn't `not` be more explicit?

Reply:
`~` is the bitwise `not`, just like `&` is the bitwise `and`. They're used with numbers or sets usually in Python, but sqlalchemy columns override them.